### PR TITLE
CUDA triangular solver

### PR DESCRIPTION
**Pull request by msimberg**
This builds on top of #315 (for fixes made to HPX since 1.5.0), and #312 (for global CUDA/cuBLAS pools/executors), so the diff is quite massive.

~Still WIP, but~ here are the main changes:
- the blas tile functions now have callable object wrappers for easier use in generic contexts
- the wrappers (together with `unwrapExtendTiles` from #312) are used in all algorithms, also the CPU-only ones
- the triangular solver is now implemented generically for CPU and GPU (but could still be specialized for a backend if needed)
- umpire is used as a memory pool for cuda allocations (both pinned host and device memory)

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -41,6 +41,9 @@ if (DLAF_WITH_CUDA)
   find_package(CUDALIBS REQUIRED)
 endif()
 
+# TODO: use option?
```
Comment (msimberg) [inline CMakeLists.txt:44]: Yes. See #317.

File: CMakeLists.txt [hunk 2]
```diff
@@ -61,13 +64,19 @@ else()
...
-find_package(HPX 1.5.0 REQUIRED)
+find_package(HPX 1.6.0 REQUIRED)
 
 # ----- BLASPP/LAPACKPP
 find_package(OpenMP REQUIRED)
 find_package(blaspp REQUIRED)
 find_package(lapackpp REQUIRED)
 
+# ----- UMPIRE
+set(DLAF_WITH_UMPIRE OFF CACHE BOOL "Enable Umpire support")
```
Comment (msimberg) [inline CMakeLists.txt:75]: Use `option`.

File: include/dlaf/communication/sync/broadcast.h [hunk 3]
```diff
@@ -54,41 +55,98 @@ void receive_from(const int broadcaster_rank, Communicator& communicator, DataOu
...
-        MemView_t mem_view(tile_size.linear_size());
-        Tile_t tile(tile_size, std::move(mem_view), tile_size.rows());
+        Tile_t tile(tile_size, MemView_t(tile_size.linear_size()), tile_size.rows());
         comm::sync::broadcast::receive_from(rank, pcomm.ref().subCommunicator(rc_comm), tile);
         return ConstTile_t(std::move(tile));
       },
       "recv_tile");
-  return hpx::dataflow(ex, std::move(recv_bcast_f), mpi_task_chain());
+
+  return detail::handle_recv_tile<D>::call(hpx::dataflow(ex, std::move(recv_bcast_f), mpi_task_chain()));
```
Comment (msimberg) [inline include/dlaf/communication/sync/broadcast.h:149]: Relevant for #318.

File: include/dlaf/tile.h [hunk 4]
```diff
@@ -0,0 +1,26 @@
+//
```
Comment (msimberg) [inline include/dlaf/tile.h:1]: Rename to `include/dlaf/blas.h`?

File: miniapp/miniapp_triangular_solver.cpp [hunk 5]
```diff
@@ -12,17 +12,23 @@
 #include <limits>
 #include <type_traits>
 
+#ifdef DLAF_WITH_CUDA
+#include <cublas_v2.h>
+#endif
```
Comment (msimberg) [inline miniapp/miniapp_triangular_solver.cpp:17]: This shouldn't be needed (anymore).

File: spack/packages/dla-future/package.py [hunk 6]
```diff
@@ -25,9 +25,10 @@ class DlaFuture(CMakePackage, CudaPackage):
     depends_on("mpi")
     depends_on("blaspp")
     depends_on("lapackpp")
+    depends_on("umpire~examples")
+    depends_on("umpire+cuda", when="+cuda")
     depends_on("hpx cxxstd=14 networking=none")
-    depends_on("hpx@1.5.0", when="~cuda")
-    depends_on("hpx@master +cuda", when="+cuda")
+    depends_on("hpx+cuda", when="+cuda")
```
Comment (msimberg) [inline spack/packages/dla-future/package.py:31]: This should be changed back to 1.6.0 when it's out.

**Comment by msimberg** [comment]
bors try

File: CMakeLists.txt [hunk 7]
```diff
@@ -23,7 +23,12 @@ if (NOT CMAKE_BUILD_TYPE)
   set(CMAKE_BUILD_TYPE "RelWithDebInfo" CACHE STRING "Build Type" FORCE)
 endif()
 
+include(CMakeDependentOption)
+
 option(BUILD_SHARED_LIBS "Build shared libraries." OFF)
+option(DLAF_WITH_CUDA "Enable CUDA support" OFF)
+cmake_dependent_option(DLAF_WITH_CUDA_MPI_RDMA "Enable MPI CUDA RDMA" OFF "DLAF_WITH_CUDA" OFF)
+cmake_dependent_option(DLAF_WITH_UMPIRE "Enable Umpire support" ON "DLAF_WITH_CUDA" OFF)
```
Comment (msimberg) [inline CMakeLists.txt:31]: Umpire is mostly useless with the CPU backend. However, it also doesn't seem to hurt, and with spack, having it as a dependency is rather trivial. My preference would be to actually make it a required dependency since it avoids having to run tests with and without it. On the other hand, I see why it might be nice to not force it as a dependency (although for CUDA it's practically a must if one wants good performance).

Preferences?

**Comment by msimberg** [comment]
bors try

**Comment by msimberg** [comment]
bors try

**Comment by msimberg** [comment]
bors try

File: CMakeLists.txt (hunk 7 above)
Comment (teonnik) [reply CMakeLists.txt:31]: My guess is that DLA-F will be built with CUDA support in production much more often than not. If that is indeed the case, I don't mind making Umpire a required dependency.

**Comment by msimberg** [comment]
bors try

**Comment by msimberg** [comment]
bors try

**Comment by msimberg** [comment]
Well, the test steps failed, but this is ready for a(nother) review now. @haampie/others is the gitlab instance known to not yet be fully functional after the upgrade?

All the tile blas operations are callable objects and should be used together with the helper `unwrapExtendTiles` from the previous PR. Umpire is still optional, but I'll change it to required if everybody thinks that's ok. The CUDA build in CI has it enabled and the CPU build has it disabled so it's decently covered, but the CUDA+no-umpire and CPU+umpire combinations are not built.

For the case when CUDA RDMA is not available, there are two helper functions `handleRecvTile` and `prepareSendTile` that copy the tiles in the GPU version. In the CPU version they are no-ops. In the terminology of https://github.com/eth-cscs/DLA-Future/pull/318#issue-555566711 these are "helpers" because of the need to specialize them into no-ops when the source and destination devices are the same. They are also now explicitly separate from the `send/recvTile` functions which will hopefully mean no/few conflicts with #228.

The `unwrapExtendTiles/getUnwrapReturnValue` combo is something that I really wish to see go away with senders/receivers (unwrapping becomes unnecessary, which in turn makes it easier for executors to generically just keep arguments alive longer if necessary), but they continue to be necessary at the moment.

Besides feedback on the functionality itself I'd appreciate comments on naming. E.g. is the helper `CommunicationDevice` confusingly named? I can see how it could be, but given that the place where memory is allocated is a `Device` it's not far from the truth.

**Comment by rasolca** [comment]
bors try
(It was a scratch cleanup problem.)

File: include/dlaf/tile.h (hunk 4 above)
Comment (albestro) [reply include/dlaf/tile.h:1]: Do you see any plus in keeping it separate? Otherwise, I would go for keeping them all together.

**Review by albestro — CHANGES_REQUESTED** [review]
In general, LGTM. Very nice work.

File: include/dlaf/communication/sync/broadcast.h [hunk 8]
```diff
@@ -55,10 +57,29 @@ void receive_from(const int broadcaster_rank, Communicator& communicator, DataOu
...
+/// Helper function for handling a tile after sending.
+///
+/// If CUDA RDMA is disabled, the tile returned from recvTile will always be on
+/// the CPU. This helper duplicates to the GPU if the first template parameter
+/// is a GPU device. The first template parameter must be given.
+template <Device D, typename T>
+auto handleRecvTile(hpx::future<matrix::Tile<T, CommunicationDevice<D>::value>> tile) {
+  return matrix::duplicateIfNeeded<D>(std::move(tile));
+}
+
```
[the remark this answers is not in the document]
Comment (msimberg) [reply include/dlaf/communication/sync/broadcast.h:78]: I agree! There's no particular reason for them to not be in `internal`. The main reason they're explicitly called inside the algorithm, rather than somewhere hidden is that the "wrapper" send/recv functions that used to call dataflow were lifted up so that dataflow is called explicitly in the algorithm. The only restriction is that `handleRecvTile/prepareSendTile` should ideally be outside the dataflow call.

`handleRecvTile` could in principle be inside `recvTile` as a `return dataflow(copyexecutor, handleRecvTile, ...)` at the end, but the future from `prepareSendTile` should really be passed to `dataflow(mpiexecutor, sendTile, ...)` instead of being called inside `sendTile`

I'm open to what you all think is clearest, but it sounds to me right now like I should reintroduce the wrapper functions that call dataflow.

File: include/dlaf/communication/sync/broadcast.h [hunk 9]
```diff
@@ -55,10 +57,29 @@ void receive_from(const int broadcaster_rank, Communicator& communicator, DataOu
...
 /// Task for broadcasting (send endpoint) a Tile in a direction over a CommunicatorGrid
-template <class T>
+template <typename T, Device D, template <class> class Future>
 void sendTile(hpx::future<common::PromiseGuard<comm::CommunicatorGrid>> mpi_task_chain, Coord rc_comm,
-              hpx::shared_future<matrix::Tile<const T, Device::CPU>> tile) {
+              Future<matrix::Tile<T, D>> tile) {
```
[the remark this answers is not in the document]
Comment (msimberg) [reply include/dlaf/communication/sync/broadcast.h:82]: It might be ok to make it `shared_future`, but I'd prefer to avoid unnecessarily converting everything to a `shared_future` if not necessary.

File: include/dlaf/matrix/copy_tile.h [hunk 10]
```diff
@@ -98,25 +99,76 @@ struct CopyTile<T, Device::GPU, Device::GPU> {
...
-void copy(TileElementSize region, TileElementIndex in_idx, const matrix::Tile<const T, Device::CPU>& in,
-          TileElementIndex out_idx, const matrix::Tile<T, Device::CPU>& out) {
+void copy(TileElementSize region, TileElementIndex in_idx, const Tile<const T, Device::CPU>& in,
+          TileElementIndex out_idx, const Tile<T, Device::CPU>& out) {
   dlaf::tile::lacpy<T>(region, in_idx, in, out_idx, out);
 }
 
 DLAF_MAKE_CALLABLE_OBJECT(copy);
+
+/// Helper struct for copying a given tile to an identical tile on Destination.
```
[the remark this answers is not in the document]
Comment (msimberg) [reply include/dlaf/matrix/copy_tile.h:125]: Yeah, identical means same type, same contents. Only the device might be different.

File: include/dlaf/solver/triangular/impl.h [hunk 11]
```diff
@@ -390,11 +381,13 @@ void Triangular<Backend::MC, Device::CPU, T>::call_LLN(comm::CommunicatorGrid gr
         auto kk = LocalTileIndex{k_local_row, k_local_col};
 
         kk_tile = mat_a.read(kk);
-        dataflow(executor_mpi, comm::sendTile_o, serial_comm(), Coord::Row, mat_a.read(kk));
+        auto comm_tile = comm::prepareSendTile(mat_a.read(kk));
```
[the remark this answers is not in the document]
Comment (msimberg) [reply include/dlaf/solver/triangular/impl.h:384]: Agree, replied in more detail on the other comment.

File: include/dlaf/tile.h (hunk 4 above)
Comment (msimberg) [reply include/dlaf/tile.h:1]: No reason other than that the blas and cublas functionality have so far stayed mostly in separate files. This is the current `cublas_tile.h`: https://github.com/eth-cscs/DLA-Future/blob/e70c6518459bf1a0159d94660156e23380a07bf8/include/dlaf/cublas_tile.h. I could move the cublas declarations to `blas_tile.h` as well, and move the cublas definitions to `blas_tile.tpp`. In that case, should things like `util_blas.h` and `util_cublas.h` also be combined?

File: miniapp/miniapp_triangular_solver.cpp [hunk 12]
```diff
@@ -41,8 +43,15 @@ using dlaf::comm::Communicator;
...
-using T = double;
+using T = float;
+#if DLAF_WITH_CUDA
+constexpr Backend B = Backend::GPU;
+using HostMatrixType = dlaf::Matrix<T, Device::CPU>;
+using MatrixType = dlaf::Matrix<T, Device::GPU>;
+#else
+constexpr Backend B = Backend::MC;
 using MatrixType = dlaf::Matrix<T, Device::CPU>;
+#endif
```
[the remark this answers is not in the document]
Comment (msimberg) [reply miniapp/miniapp_triangular_solver.cpp:54]: I think it should eventually just be a command line option to the miniapps.

File: spack/packages/dla-future/package.py [hunk 13]
```diff
@@ -25,6 +25,8 @@ class DlaFuture(CMakePackage, CudaPackage):
     depends_on("mpi")
     depends_on("blaspp")
     depends_on("lapackpp")
+    depends_on("umpire~examples")
+    depends_on("umpire+cuda~shared", when="+cuda")
```
[the remark this answers is not in the document]
Comment (msimberg) [reply spack/packages/dla-future/package.py:29]: No. It should be disabled for cuda as well with that specification unless I'm missing something.

**Comment by msimberg** [comment]
bors try

File: spack/packages/dla-future/package.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (albestro) [reply spack/packages/dla-future/package.py]: my fault, you're right!

File: miniapp/miniapp_triangular_solver.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (msimberg) [reply miniapp/miniapp_triangular_solver.cpp:54]: I suppose this could also be handled in a separate PR?

**Comment by msimberg** [comment]
@albestro I've addressed some of your comments. I didn't yet do anything about
- the communication helpers (I moved them to `internal` but did not move the calls)
- did not move/merge any header files
I'd appreciate yours (and others') feedback on those before I do anything about them.

File: include/dlaf/communication/sync/broadcast.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (albestro) [reply include/dlaf/communication/sync/broadcast.h:84]: Can I ask you the rationale? Because usually I always tried to ensure const-correctness, while it seems there is a more important reason for not doing so.

[the remark this answers is not in the document]
Comment (msimberg) [reply include/dlaf/communication/sync/broadcast.h]: After discussing with @albestro I understood that I had read his question sloppily. I understood it to be about the change to have a generic `Future` type, rather than about the constness of `T`. It can and should indeed be `const` and I've just pushed a change which makes it `const`. Thanks @albestro for pointing this out!.

**Comment by msimberg** [comment]
bors try

File: miniapp/miniapp_triangular_solver.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (rasolca) [reply miniapp/miniapp_triangular_solver.cpp:54]: I agree to have an option for it and that it can be handled in a separate PR.

File: include/dlaf/tile.h (hunk 4 above)
Comment (rasolca) [reply include/dlaf/tile.h:1]: What about merging `include/dlaf/tile.h``include/dlaf/blas_tile.h` and `include/dlaf/cublas_tile.h` in a single file `include/dlaf/blas/tile.h`?
I would anyway keep `util_blas.h` and `util_cublas.h` separate and move them in `include/dlaf/blas` as well.

File: include/dlaf/matrix/tile.h [hunk 14]
```diff
@@ -222,11 +223,9 @@ struct UnwrapFuture<hpx::future<Tile<T, D>>> {
 
 /// Callable object used for the unwrapExtendTiles function below.
 template <typename F>
-struct UnwrapExtendTiles {
-  F f;
-
+class UnwrapExtendTiles {
   template <typename... Ts>
-  auto operator()(Ts&&... ts) {
+  auto call_helper(std::true_type, Ts&&... ts) {
```
Comment (rasolca) [inline include/dlaf/matrix/tile.h:228]: According to the style:
```suggestion
  auto callHelper(std::true_type, Ts&&... ts) {
```

**Review by rasolca — APPROVED** [review]
Looks good.
Just spotted a minor naming issue.

File: include/dlaf/tile.h (hunk 4 above)
Comment (msimberg) [reply include/dlaf/tile.h:1]: > What about merging `include/dlaf/tile.h`, `include/dlaf/blas_tile.h`, `include/dlaf/blas_tile.tpp` and `include/dlaf/cublas_tile.h` in a single file `include/dlaf/blas/tile.h`?

Ok with me. @albestro?

> I would anyway keep `util_blas.h` and `util_cublas.h` separate and move them in `include/dlaf/blas` as well.

Separately into `dlaf/blas/util.h` and `dlaf/cublas/util.h`?

I suppose there are related renamings that could be done for `lapack_tile.h`, `util_math.h`, `util_matrix.h`, and `util_tile.h`, but I won't move those in this PR.

File: CMakeLists.txt (hunk 7 above)
Comment (msimberg) [reply CMakeLists.txt:33]: I'll make it required in that case.

File: miniapp/miniapp_triangular_solver.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (msimberg) [reply miniapp/miniapp_triangular_solver.cpp:54]: Opened https://github.com/eth-cscs/DLA-Future/issues/345 as a reminder to change it afterwards.

File: include/dlaf/communication/sync/broadcast.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (albestro) [reply include/dlaf/communication/sync/broadcast.h]: After your offline explanation of the restrictions I see your point.

IMHO in general we should try to prefer "schedulable" functions instead of "scheduling" ones (for reference see https://github.com/eth-cscs/DLA-Future/pull/318#issuecomment-767350611).

I think that with the new needs  of `handle/prepare` for communications, plus the fact that we always want to use the same executor for communication (right?), it is reasonable to make them "scheduling" functions, even if we will have to investigate if there is an easy way to customize annotation for them (e.g. "send diagonal tile row-wise" instead of a generic "send tile").

@msimberg @rasolca @teonnik from an API point of view, should we make it clear that they inside call `dataflow`, so they shall not go into a task?

**TL;DR** I agree you need to reintroduce the wrapper functions that calls dataflow (what I refer to as "scheduling" functions).

**Comment by msimberg** [comment]
bors try

**Comment by msimberg** [comment]
bors try

File: include/dlaf/communication/sync/broadcast.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (msimberg) [reply include/dlaf/communication/sync/broadcast.h]: Well, I renamed the wrappers to `scheduleSendTile`/`scheduleRecv(Alloc)Tile` now. Alternatively I'd simply call them `sendTileWrapper` etc. I'm sure we'll come up with a better naming scheme in the future. Some potential alternatives:

- Move the "task" functions to the `internal` namespace to indicate that they shouldn't be called directly and call the wrappers plain `recv/sendTile`. I didn't move them now since they are still used in non-GPUized algorithms. The fact that the wrappers take an executor as the first argument is a good indication that they shouldn't be used directly as tasks.
- Somehow indicate the RDMA-ness of the communication functions. This is in principle orthogonal to whether or not they should be tasks in themselves, but may help clarify how they should be used. E.g. something like `recvTileRDMA` for the plain version that always receives on the target device and `recvTileViaCPU` for  a version that receives via the CPU (conditionally or not?).

Unless someone gets a big a-ha moment from these ideas I suggest we go ahead with something relatively generic like the `schedule*` or `*Wrapper` names.

**Comment by msimberg** [comment]
bors try

**Comment by msimberg** [comment]
bors try

**Review by teonnik — APPROVED** [review]

File: include/dlaf/init.h [hunk 15]
```diff
@@ -48,6 +50,12 @@ struct Init {
   static void finalize() {}
 };
 
+template <>
+void Init<Backend::MC>::initialize(configuration const&);
+template <>
+void Init<Backend::MC>::finalize();
+extern template struct Init<Backend::MC>;
+
```
[the remark this answers is not in the document]
Comment (teonnik) [reply include/dlaf/init.h:58]: I forgot to add a comment, so I  will do it here: 

LGTM overall. 

The triangular solver has to be adapted to the new MPI API. The implementation of the MPI executor allows to be compatible with `getMPIExecutor<>()` API.

[the remark this answers is not in the document]
Comment (msimberg) [reply include/dlaf/init.h:58]: > Do you need this in the header for testing purposes?

Good question, I don't remember (it wasn't for testing purposes as far as I can remember). It'd indeed be nice if it can go in the cpp file. Will check again.

> I forgot to add a comment, so I will do it here:
> 
> LGTM overall.
> 
> The triangular solver has to be adapted to the new MPI API. The implementation of the MPI executor allows to be compatible with `getMPIExecutor<>()` API.

Yeah, I'm changing that right now. Given that the executor's constructor now chooses the correct pool `getMPIExecutor` is somewhat unnecessary. It might be worth keeping it just for symmetry with the other executors though.

Did I understand it correctly that one doesn't actually need a separate row and column executor (they're exactly the same)? I presume this was from trying to have the executor know which type it is (which we should still look at)?

[the remark this answers is not in the document]
Comment (teonnik) [reply include/dlaf/init.h:58]: > Did I understand it correctly that one doesn't actually need a separate row and column executor (they're exactly the same)? I presume this was from trying to have the executor know which type it is (which we should still look at)?

Yes, the row/column distinction comes from the communicator which is now handled outside of the executor. The only exception is if we want to later bring back "blocking" MPI version of the executor. In that case we would need to save the index of the OS thread to which the executor is assigned. That means we would need two different executors for the row and column communicators. However, as we decided to remove the "blocking" version, a single executor is sufficient.

File: spack/packages/dla-future/package.py [hunk 16]
```diff
@@ -25,7 +25,9 @@ class DlaFuture(CMakePackage, CudaPackage):
     depends_on("mpi")
     depends_on("blaspp")
     depends_on("lapackpp")
-    depends_on("hpx cxxstd=14 networking=none")
+    depends_on("umpire~examples")
+    depends_on("umpire+cuda~shared", when="+cuda")
+    depends_on("hpx cxxstd=14 networking=none +async_mpi")
```
Comment (msimberg) [inline spack/packages/dla-future/package.py:30]: NB. I added `+async_mpi` here. If someone wants it I can make a separate PR for that.

**Comment by msimberg** [comment]
bors try

File: include/dlaf/init.h (hunk 15 above)
[the remark this answers is not in the document]
Comment (msimberg) [reply include/dlaf/init.h]: > Do you need this in the header for testing purposes?

It was unnecessary, I've moved it to the cpp file.

For the triangular solver I now just use one executor for both column and row communication. It should be easy enough to change it back when/if needed.

**Comment by msimberg** [comment]
This seems to pass tests locally now after the merge with master.

I made one additional change to the header file organization: I moved the triangular solver declarations required by `triangular.h` to `triangular/api.h`, and `triangular/impl.h` is now never included in user-code, since it's not necessary (only necessary for the instantiations). This means that changes to the triangular solver implementation doesn't cause tests/examples using it to be recompiled. Let me know if you think this sounds reasonable, or if I should revert and we handle this separately.

File: include/dlaf/init.h (hunk 15 above)
[the remark this answers is not in the document]
Comment (albestro) [reply include/dlaf/init.h]: I agree with you about having a single Executor for both row and col...

I have one more question: why not having a single Executor also for all MPI communications? I mean, for the moment is not a big deal because it does not take any parameter, but it may be reasonable to have it inside our `initialize` and get it via the `getMPIExecutor()`. Do you see any problem? This may be the topic of a dedicated issue/PR anyway...

[the remark this answers is not in the document]
Comment (msimberg) [reply include/dlaf/init.h]: My wording was maybe confusing. The same executor is used for both types of communication now. If you're talking about using literally the same instance that's also a possibility (i.e. returning a reference to a function static executor); currently it's a new instance every time you call `getMPIExecutor`, but I don't see that as a huge overhead (executors are cheap to create).

**Comment by msimberg** [comment]
bors try

**Comment by msimberg** [comment]
bors try

**Comment by msimberg** [comment]
bors try

**Comment by rasolca** [comment]
bors try
(codecov checksum should be fixed)

**Comment by msimberg** [comment]
bors try

**Comment by msimberg** [comment]
The one failure looks to me unrelated to the changes here: https://gitlab.com/cscs-ci/eth-cscs/DLA-Future/-/jobs/1202885692. It seems to have successfully run the tests but was stuck on creating the coverage reports.

**Comment by rasolca** [comment]
It is not stuck, but it takes long time to initialize the report and to create it and we often hit the time limit.
I'll check if I find a way to speed things up, otherwise I'll increase the time limit.

**Comment by msimberg** [comment]
bors try

**Review by albestro — APPROVED** [review]
LGTM
